NetKVM/Common/ParaNdis_TX.cpp

```diff
@@ -1151,3 +1193,76 @@ ULONG CNB::Copy(PVOID Dst, ULONG Length, ULONG DataOffset) const
 
     return Copied;
 }
+
+void CNB::ReturnPages()
+{
+    if (m_ExtraNBStorage)
+    {
+        m_ParentNBL->GetParentTXPath()->ReturnPages(m_ExtraNBStorage);
+        NdisFreeMemory(m_ExtraNBStorage, sizeof(CExtendedNBStorage), 0);
+    }
+    m_ExtraNBStorage = nullptr;
+}
+
+NBMappingStatus CNB::AllocateAndFillCopySGL()
+{
+    //decide how many pages should be used 
+    ULONG uDataLength = GetDataLength();
+    ULONG uPages = (GetDataLength() + PAGE_SIZE - 1) / PAGE_SIZE;
+
+    m_ExtraNBStorage = (CExtendedNBStorage*)ParaNdis_AllocateMemory(m_Context, sizeof(CExtendedNBStorage));
+
+    if (m_ExtraNBStorage == nullptr)
+    {
+        DPrintf(0, "[%s] ExtendedNBStorage allocation failed \n", __FUNCTION__);
+        return NBMappingStatus::FAILURE;
+    }
+
+    RtlZeroMemory(m_ExtraNBStorage, sizeof(CExtendedNBStorage));
+    if (!m_ParentNBL->GetParentTXPath()->BorrowPages(m_ExtraNBStorage, uPages))
+    {
+        return NBMappingStatus::NO_RESOURCE;
+    }
+
+    ULONG uDataOffset = 0;
+    for (ULONG i = 0; i < uPages; i++)
+    {
+        PVOID pVirtualAddress = m_ExtraNBStorage->m_UsedPages[i]->GetVA();
+        ULONG uCopySize = (i == uPages - 1) ? (uDataLength - uDataOffset) : PAGE_SIZE;
+        ULONG uCopied = Copy(pVirtualAddress, uCopySize, uDataOffset);
+        if (uCopied != uCopySize)
+        {
+            ReturnPages();
+            DPrintf(0, "[%s] copy failed! expected %lu, copied %lu bytes\n", __FUNCTION__, uCopySize, uCopied);
+            return NBMappingStatus::FAILURE;
+        }
+        m_ExtraNBStorage->m_Elements[i].Address = m_ExtraNBStorage->m_UsedPages[i]->GetPA();
+        m_ExtraNBStorage->m_Elements[i].Length = uCopySize;
+        uDataOffset += uCopySize;  // Adjust offset, last page might not need adjustment but does not harm
+    }
+    return NBMappingStatus::SUCCESS;
+}
+
+NBMappingStatus CNB::MapCopyDataToVirtioSGL(CTXDescriptor &Descriptor, ULONG Offset) const
+{
+    for (ULONG i = 0; i < m_ExtraNBStorage->m_UsedPagesCount; i++)
+    {
+        if (Offset < m_ExtraNBStorage->m_Elements[i].Length)
+        {
+            PHYSICAL_ADDRESS PA;
+            PA.QuadPart = m_ExtraNBStorage->m_Elements[i].Address.QuadPart + Offset;
+
+            if (!Descriptor.AddDataChunk(PA, m_ExtraNBStorage->m_Elements[i].Length - Offset))
+            {
+                return NBMappingStatus::FAILURE;
+            }
+            Offset = 0;
```

Comment:
Sorry, I was mistaken in my previous comment, I've removed it.
Let me check better, I'll do it today later.